Core/Src/main.c

```diff
@@ -87,14 +89,39 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
-
+  __HAL_TIM_SET_COMPARE(&htim1, TIM_CHANNEL_1, 0.05 * __HAL_TIM_GET_AUTORELOAD(&htim1));
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+    /* pot on CH0 of MCP3004 (single input), send bits [1 1 x 0 0] (start, sgl/diff, d2, d1, d0)*/
+    uint8_t tx_data[3] = {0x1, 0x1 << 7, 0x0}; /* pad with zeroes to fit 8 bits */
+    uint8_t rx_data[3] = {0};
+
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET); /* CS low (start transmit) */
+
+    /* full duplex mode, use transmit+receive */
+    HAL_SPI_TransmitReceive(&hspi1, tx_data, rx_data, 3, HAL_MAX_DELAY);
+
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET); /* CS high (end transmit) */
+
+    /* get last 10 out of 24 bits received from ADC */
+    uint16_t adc_output = ((rx_data[1] & 0x03) << 8) | rx_data[2];
+
+    /* max ADC output is 2^10 - 1 = 1023 */
+    /* 5-10% duty cycle with period = 60000, so values range from 3000-6000 */
+    uint16_t adc_compare = (1 + (adc_output / 1023)) * 0.05 * __HAL_TIM_GET_AUTORELOAD(&htim1);
```

Comment:
`(adc_output / 1023)` would result in a whole number instead of a floating point number since two operands are whole numbers. This will lead to incorrect results.